### DISCUSSION TITLE
Severe npm audit vulnerability - how can I fix this?

### DISCUSSION THREAD

**Discussion by Svarto** [gatsbyjs/gatsby#28852] [Help]
Hi Gatsby team and community,

Apologies if this is a silly question, but I don’t know how to fix it and hope you can help.

When installing the latest version of Gatsby I get 2 severe npm audit warnings. They refer to an old version of Axios that is used by Gatsby, is there somehow I can update this in the Gatsby package or fix them? The command npm audit fix does not work.

My package.json:

"dependencies": {
    "autoprefixer": "^10.1.0",
    "gatsby": "^2.29.3",
    "gatsby-image": "^2.8.0",
    "gatsby-plugin-google-analytics": "^2.8.0",
    "gatsby-plugin-postcss": "^3.4.0",
    "gatsby-plugin-react-helmet": "^3.7.0",
    "gatsby-plugin-sharp": "^2.11.2",
    "gatsby-plugin-sitemap": "^2.9.0",
    "gatsby-source-filesystem": "^2.8.1",
    "gatsby-transformer-sharp": "^2.9.0",
    "postcss": "^8.2.2",
    "prop-types": "^15.7.2",
    "react": "^17.0.1",
    "react-dom": "^17.0.1",
    "react-helmet": "^6.1.0",
    "sharp": "^0.27.0",
    "tailwindcss": "^2.0.2"
  },

The npm audit security warnings:
![image](https://user-images.githubusercontent.com/24279217/103628353-c82aab80-4f36-11eb-8dca-08ffb92ce88b.png)

**Comment by nikoladev** [comment]
It seems all Gatsby packages using `axios` should be updated to use `0.21.1` at least.

As for the `axios` version pulled in by `contentful-management`, [from `contentful-management@6.3.1` onwards](https://github.com/contentful/contentful-management.js/pull/514#issuecomment-728898001) it uses `axios@^0.21.0` (which will allow users to patch `axios` to `0.21.1` themselves). So `gatsby-recipes` can use `contentful-management@^6.3.1` (or even `contentful-management@^7`)

@Svarto As for what you can do now: I don't know if npm has something like this, but yarn has [yarn resolutions](https://classic.yarnpkg.com/en/docs/selective-version-resolutions/) which can force `axios` to version `0.21.1`.

Edit: it seems that [this PR](https://github.com/gatsbyjs/gatsby/pull/28545) will fix the vulnerabilities because it updates both `axios` and `contentful-management`

**Comment by Joshua-Flores** [comment]
Same here, wondering what the deal is to fix this.

**Comment by janaagaard75** [comment]
I think we have to speak up, somehow, to get Gatsby to focus on this. I believe that most of the issues could be solved by Gatsby updating their dependencies, and that in most cases making the change is simply updating the version number. In reality the task is a lot bigger though, since any changes rightfully requires a lot of testing, but that could probably be done by releasing a version with the updates as beta, to get the community to help verifying that everything still works.

**Comment by dan-kirkham** [comment]
@janaagaard75 There's nothing stopping someone opening a PR if it's meaningful enough to you. I've done this before several times and the core Gatsby team were happy to merge these. The reality is that npm audit doesn't understand how axios is used, so it can only show a high warning, but in reality _you_ own the build machine that runs axios, so the threat is minimal.

There's so many high priority things that the core team should be focusing on and I can't see why the community can't pick up these more trivial changes.

**Comment by LekoArts** [comment]
While those audit messages might seem scary they are almost always irrelevant in the context of Gatsby as it's a build tool. So there is nothing to "fix" as the warning is not relevant.

Gatsby creates static assets and runs everything at build time and not during runtime. `npm audit` is designed for runtime / Node apps so it flags issues that can occur there. This means that almost every "vulnerability" report we receive are false positives. While in principle Gatsby can also have vulnerabilities you need to make sure that it's relevant to Gatsby before reporting it. For example a "Regex DDOS attack" can never be a real vulnerability for a development-time tool. If you want to override/update a transitive dependency you can use [yarn resolutions](https://classic.yarnpkg.com/en/docs/selective-version-resolutions/).

You can also read [npm audit: Broken by Design](https://overreacted.io/npm-audit-broken-by-design/) to learn more about this.

If you know that a vulnerability affects Gatsby because you understand what the vulnerability is, please report it [here](https://github.com/gatsbyjs/gatsby/blob/master/SECURITY.md). Thanks!
